### DISCUSSION TITLE
Issue with File Upload to S3 using Filament in Laravel

### DISCUSSION THREAD

**Discussion by evan8133** [filamentphp/filament#11021] [Help] [answered]
Hello everyone,

I'm relatively new to Filament, and I'm encountering an issue with uploading files to Amazon S3 using Filament. Strangely, my Laravel code for file uploads works perfectly fine. I have followed the documentation closely, but I seem to be missing something. Here's how I achieve file uploads in Laravel:
```
$profilePicture = $request->file('profile_picture');
$profilePictureName = time() . '.' . $profilePicture->extension();
$file = $profilePicture->storeAs('profile_pictures', $profilePictureName, 's3');
$url = Storage::disk('s3')->url($file);
$userModel->avatar = $url;
$userModel->save();
return response()->json([
    'message' => 'Profile picture updated successfully',
    'user' => StudentResource::make($userModel),
]);
```
Now, here's my Filament code for file uploads:

```
Forms\Components\FileUpload::make('file')
    ->required()
    ->label('File')
    ->maxSize(10240)
    ->maxFiles(1)
    ->disk('s3')
    ->directory('handbooks')
    ->visibility('public')
    ->placeholder('Upload File'),
```
When I try to upload a file using Filament, I receive a success message, but the file is not being uploaded to S3. Additionally, the file path and name returned are not as expected (e.g., handbook\'filename'). In contrast, my Laravel code provides the public URL for the uploaded file.

I have thoroughly checked the documentation but couldn't identify the issue. Any insights or suggestions on how to resolve this would be greatly appreciated.

Thank you!

**Comment by elmudometal** [comment]
I have the same problem, I solved it by putting visibility = "publico" because investigating is a problem in BaseFileUpload.php and HasFileAttachments.php that asks:
$storeMethod = $component->getVisibility() === 'public' ? 'storePubliclyAs' : 'storeAs';

and for some reason that I don't understand storePubliclyAs is not saving the image

**Reply by evan8133** [reply]
thanks for the replay, are u saying to chain this  ->visibility('public') because i already did that or do it as you wrote it  ->visibility('publico')

**Reply by elmudometal** [reply, marked as answer]
I used literal ->visibility('publico') to prevent it from doing storePubliclyAs

you can try to see if it works for you

**Reply by evan8133** [reply]
Well im really surprised that it worked this way, your a life saver man, by the way to you know to to get the file URL instate of the path this is how i do it:
```
Forms\Components\FileUpload::make('file')
                            ->required()
                            ->label('File')
                            ->maxSize(10240)
                            ->maxFiles(1)
                            ->disk('s3')
                            ->directory('handbooks')
                            ->visibility('publico')
                            ->placeholder('Upload File'),
```

**Reply by elmudometal** [reply]
What route do you want to get?
because the path would be storage::disk("s3")->url($data->file)

**Reply by evan8133** [reply]
I want to get the public URL of the file when uploaded and store that in the database , if possible good if not is fine, the fact now it work is perfect and thank you so much

**Reply by elmudometal** [reply]
The ideal is that you save the relative path and with the Storage helper you see it in your public view, so if you change any configuration it will take care of resetting everything without additional work

**Reply by elmudometal** [reply]
If you don't want to have to use every moment you can create an attribute in the model that returns the full url

**Reply by mokalovesoulmate** [reply]
> I used literal ->visibility('publico') to prevent it from doing storePubliclyAs
> 
> you can try to see if it works for you

Hello @elmudometal , from where did you get this `publico` thing? I am currently solving a problem where Filament cannot upload to an S3 that has "Block public access" activated because it is required by the administrator. 

This `publico` is magically solving the problem. The file now can be uploaded.

**Reply by elmudometal** [reply]
After reviewing the component code I noticed that the problem is with the storePubliclyAs function, this does not work correctly, therefore to move on to the next method without generating an error, use 'publico' visibility and with that jump to the storeAs method and use the configuration that you set in your AWS without problem.

**Reply by chrissy-dev** [reply]
@elmudometal, thank you for this. This solved my issue for now. I'm hoping we can get a core fix in place for this if it isn't desired behaviour. If it is, it should be documented. 

@mokalovesoulmate As far as I can see from the component code mentioned, "`publico`" isn't specifically solving the problem. You could put anything as the `visibility->('abcd')` value except `public` and it'll work as it'll fall back to `storeAs`. 

There's an issue about this since Dec 2023. @danharrin I think @elmudometal has identified the issue in https://github.com/filamentphp/filament/issues/10379#issuecomment-1861517170.

💘

**Reply by danharrin** [reply]
By default, we store files publicly as its a safer default, not all drivers (such as the default local driver) support generating temporary URLs which are required for storing files privately. As such this default shall not change unless a new configuration option is introduced.

Please refer to the docs where we specifically have an example using `private` as the visibility as part of the disk config - https://filamentphp.com/docs/3.x/forms/fields/file-upload#configuring-the-storage-disk-and-directory

This may be applied to all file upload fields at once using global settings, meaning a new config option doesn't necessarily need to be introduced. https://filamentphp.com/docs/3.x/forms/fields/getting-started#global-settings

**Reply by buddhaCode** [reply]
I stumbled around this issue, too. And I can't get my head around this.

I have a public S3 bucket (with CORS headers set) and want to upload directly to this bucket without touching the server with a temp file upload, because I am in a server-less environment. Why should I need to set the visibility to private? And why is the file not private after the upload?! I think this API is confusing.

Also, the example in Dan's first link doesn't explain what `->visibility('private')` actually does and what the whole example is about.

**Reply by danharrin** [reply]
Why do you need to set the visibility to private if its a public bucket? If its public, we dont need to use `temporaryUrl()` to generate URLs for example

**Reply by buddhaCode** [reply]
Yes, that's why I am confused ;)

I need the temporary url for the upload, not to display a preview image of the upload. I guess this is pretty much an edge case.

![image](https://github.com/user-attachments/assets/96e94f18-52cc-49c3-be78-f60065b8b724)

I can see the following behavior:

**Settings**
- File Upload Component: `FileUpload::make('my_image')->directory('my_dir')`
- .env:
```
FILESYSTEM_DISK=s3
FILAMENT_FILESYSTEM_DISK=s3
```
With this settings the default value `public` will be applied for the visibility.

**Steps**
- Select an image on the input
- See the image gets uploaded
- See the correct preview image
- See that the image got upload to `livewire-tmp` in the S3 bucket
- Save the form
- See that the temporary upload in `livewire-tmp` got deleted
- See no new file in `my_dir`
- I can not access the file via a public URL because it does not exist

**Expected Behavior**
The upload file should be saved to `my_dir` but it doesn't.

Let's try what @elmudometal proposed:

**Settings**
- File Upload Component: `FileUpload::make('my_image')->directory('my_dir')->**visibility('private')**`
- .env:
```
FILESYSTEM_DISK=s3
FILAMENT_FILESYSTEM_DISK=s3
```

**Steps**
- Select an image on the input
- See the image gets uploaded
- See the correct preview image
- See that the image got upload to `livewire-tmp` in the S3 bucket
- Save the form
- See that the temporary upload in `livewire-tmp` got deleted
- **See new file in `my_dir`**
- I **can** access the file via a public URL.

**Expected Behavior**
It does what I want. The file got uploaded to the desired folder on my S3 bucket.
But now I am confused. Although the whole S3 bucket is public, I would expect that the access rights for the particular file got overwritten and the file is not public anymore. Because its visibility is set to private.
Maybe the naming of the `visibility` method is not self explanatory. This method's only mention in the docs is in the Infolist ImageEntry component. And this chapter is about temporary urls. Which is maybe what the whole thing is about?! It's about using a temporary URL and not about setting or applying access rights to the file after it got uploaded.

**Reply by danharrin** [reply]
Sounds like the bug here is that `storePubliclyAs()`, which is whats used when you dont specify a visibility to move the file from the temp directory to S3, does not work with your bucket. That file is a feature of Laravel's `UploadedFile` class.

**Reply by buddhaCode** [reply]
To be honest. I don't know. I spend the last 2 hours source diving `\Livewire\Features\SupportFileUploads\TemporaryUploadedFile`, `\Illuminate\Http\UploadedFile` and some of the related classes in the filesystem adapter. I couldn't find anything that helps. This is way too complicated for me.
In the end, I don't care so much about the file upload. I want to use the RichEditor with an image. I just took the simpler file upload to debug this issue.
But setting the visibility to private in the RichEditor results in having all the signed image URLs in the database. That's bad, because they expire after 5 minutes. I will simply patch the RichEditor to use `storeAs()` instead of `storeAsPubliclyAs()` regardless of the visibility.
That solves my problem. But it doesn't feel right.

**Reply by elmudometal** [reply]
> Sounds like the bug here is that `storePubliclyAs()`, which is whats used when you dont specify a visibility to move the file from the temp directory to S3, does not work with your bucket. That file is a feature of Laravel's `UploadedFile` class.

For this reason I used ->visibility('publico') instead of public so that it doesn't use this method storePubliclyAs and jumps to the other one but at the same time it is not private

**Reply by danharrin** [reply]
If someone opens an issue with a reproduction repo, it'll stay on our radar and eventually someone will investigate a better fix. Discussions are not tracked as bug reports.

**Reply by buddhaCode** [reply]
I think I have found the cause of my problem. I did a deep dive into the Flysystem S3 adapter. I found out that the visibility of a file is based on the ACL of the ~bucket/~ file.
My bucket had ACLs turned off. So I turned it on ~and applied a read grant to everyone~.
With that set, almost everything works fine. I don't need to set `visibility()` to anything weird. Just use the default `public` setting.
Only one thing is odd. The uploaded file has been copied to the desired directory. But the temporary upload file is still in the `/livewire-tmp/` folder. So it was copied but not moved. This is a small problem for me. The tmp folder is cleaned up once a day anyway.
The S3 configuration seems to be the problem. I haven't found any good documentation or tutorials on how to do a proper and safe S3 configuration. I'm not entirely happy with enabling ACL, as AWS puts a lot of explanatory marks around this feature and recommends that this feature shouldn't be used anymore.

**Reply by wychoong** [reply]
I have encountered a similar issue with RichEditor and S3 as the disk with ACL turned off and serve files with Cloudfront.
the issue seems to be when ACL is off, it is not able to upload files with visibility = public
so when set `->fileAttachmentsVisibility('private')` the upload is success but it comes with second issue where the url returned to RichEditor is a temporary url which will expire after 5 mins.

if we look at Spatie's Media Library (via filament plugin), it is stored using the filesystem `put` with visibility part of the custom header

**tldr**: S3 with ACL off not able to use `storePublicly` and signed temporary url is stored in database rendered the image upload in richEditor useless

my workaround: 
`fileAttachmentsVisibility('private')`
override `handleUploadedAttachmentUrlRetrieval` with `getUploadedAttachmentUrlUsing` and commeted these lines
https://github.com/filamentphp/filament/blob/3b25f678cbaebb7e68bb895d2df8c22822ba15a1/packages/forms/src/Components/Concerns/HasFileAttachments.php#L120-L129
